Pkeller/memmap profiler

mo-tenstorrent opened the pull request:
### Ticket
#4984 

### Problem description
Profiler is using the old L1 addressing scheme

### What's changed
Make profiler use the mailbox struct to share its address between device and host

### Checklist
- [ ] Post commit CI passes: 
- [x] T3K Profiler: https://github.com/tenstorrent/tt-metal/actions/runs/10819716295
- [x] Device Perf: https://github.com/tenstorrent/tt-metal/actions/runs/10819718173
- [x] uBenchmark: https://github.com/tenstorrent/tt-metal/actions/runs/10617765919
- [x] BH Post commit: https://github.com/tenstorrent/tt-metal/actions/runs/10819693142
- [x] T3K nightly : https://github.com/tenstorrent/tt-metal/actions/runs/10819734572
- [x] T3K unit tests: https://github.com/tenstorrent/tt-metal/actions/runs/10819710955
- [x] Model's perf: https://github.com/tenstorrent/tt-metal/actions/runs/10819722464

pgkeller committed:
#4984: Move profiler buffers to mailboxes

And out of the common runtime address map

pgkeller committed:
#4984: Update storage core addr to mailbox base

Storage cores used to include mailboxes for the L1 barrier
L1 barrier is no longer in the mailboxes, so update the base
address to the base of the mailboxes

mo-tenstorrent committed:
#4984: Use globals for data and control arrays

mo-tenstorrent committed:
#4984: Fix addressing issues on eth cores

abhullar-tt commented on tt_metal/hostdevcommon/common_runtime_address_map.h:
```diff
@@ -20,58 +21,31 @@ constexpr static std::uint32_t DRAM_ALIGNMENT = NOC_DRAM_READ_ALIGNMENT_BYTES >=
+// TODO: move these out of the memory map into profiler code
 constexpr static std::uint32_t PROFILER_OP_SUPPORT_COUNT = 1000;
-constexpr static std::uint32_t PROFILER_FULL_HOST_VECTOR_SIZE_PER_RISC = PROFILER_L1_MARKER_UINT32_SIZE * (PROFILER_L1_PROGRAM_ID_COUNT +  PROFILER_L1_GUARANTEED_MARKER_COUNT + PROFILER_L1_OP_MIN_OPTIONAL_MARKER_COUNT) * PROFILER_OP_SUPPORT_COUNT;
+constexpr static std::uint32_t PROFILER_FULL_HOST_VECTOR_SIZE_PER_RISC = kernel_profiler::PROFILER_L1_MARKER_UINT32_SIZE * (kernel_profiler::PROFILER_L1_PROGRAM_ID_COUNT +  kernel_profiler::PROFILER_L1_GUARANTEED_MARKER_COUNT + kernel_profiler::PROFILER_L1_OP_MIN_OPTIONAL_MARKER_COUNT) * PROFILER_OP_SUPPORT_COUNT;
```
does `PROFILER_FULL_HOST_VECTOR_SIZE_PER_RISC` need to be in here? is it only used by profiling code? likewise for `PROFILER_FULL_HOST_BUFFER_SIZE_PER_RISC`

abhullar-tt commented on tt_metal/hostdevcommon/profiler_common.h:
```diff
@@ -24,18 +24,16 @@ namespace kernel_profiler{
     enum ControlBuffer
     {
-        HOST_BUFFER_END_INDEX_BR,
+        HOST_BUFFER_END_INDEX_BR_ER,
```
is this because eth cores treated as if there is one brisc? this may not be scalable for BH which has 2 riscvs on eth cores

mo-tenstorrent replied:
Good call, this will turn into a union and each device can define its own struct for controls.

abhullar-tt commented on tt_metal/hw/inc/blackhole/dev_mem_map.h:
```diff
@@ -53,7 +53,8 @@
 #define MEM_MAILBOX_BASE 16
 // Magic size must be big enough to hold dev_msgs_t.  static_asserts will fire if this is too small
-#define MEM_MAILBOX_END (MEM_MAILBOX_BASE + 1356)
+#define MEM_MAILBOX_SIZE 5 * 4 * 512 + 4 * 32 + 1364
```
comments for constants?

abhullar-tt commented on tt_metal/hw/inc/blackhole/eth_l1_address_map.h:
```diff
@@ -60,16 +60,9 @@ struct address_map {
-
-  static_assert((PROFILER_L1_BUFFER_ER % 32) == 0);
-  static_assert((PROFILER_L1_BUFFER_CONTROL % 32) == 0);
+  static constexpr std::uint32_t ERISC_MEM_MAILBOX_END = ERISC_MEM_MAILBOX_BASE + 288 + 256 + 16 + (32 + 512) * 4;
```
comments for constants?

mo-tenstorrent commented:
@abhullar-tt the comments are all spot on. They all relate to what is coming next. Plan was to follow up this PR with switching the Buffers to structs. 

For the size magic numbers you mentioned, the will be extracted form struct size and many of the constexpr will be removed. With the remaining ones staying in profiler code. 

I am starting to think maybe I should just switch to structs in the same PR.

abhullar-tt commented:
> @abhullar-tt the comments are all spot on. They all relate to what is coming next. The plan was to follow up this PR with switching the buffers to structs.
> 
> For the size magic numbers you mentioned, they will be extracted form struct sizes and many of the constexpr will be removed. With the remaining ones staying in profiler code.
> 
> I am starting to think maybe I should just switch to structs in the same PR and address all these issues.

ah okay, that makes sense. I think its okay to do this in a follow up PR. Can you also run BH post commit?

tt-asaigal commented on tt_metal/hw/inc/dev_msgs.h:
```diff
@@ -234,13 +229,34 @@ struct watcher_msg_t {
     struct dprint_buf_msg_t dprint_buf;
+    uint32_t pad1;
+    uint32_t pad2;
+    struct profiler_msg_t profiler;
```
Is the padding here to ensure that the profiler struct base address respects NOC alignment?

mo-tenstorrent replied:
That is right.

pgkeller replied:
the downside of an array of pads is as we add/remove members the array size could be 0 which won't compile

mo-tenstorrent replied:
The size is chosen in the ifdef depending ERISC vs TENSIX. It is dependant on that because of the variable size of DPRINT and PROFILER.  I double checked and compile time size of zero is fine, it compiles.

mo-tenstorrent commented:
As per conversation with @abhullar-tt different padding sizes depending on the device (BH vs GS/WH) needs to be added to the mailbox struct to be PICe/DRAM aligned. 

Different size padding can come in similar to https://github.com/tenstorrent/tt-metal/blob/84f944d1f5e733f0ab060e62600574a4aaefa279/tt_metal/impl/dispatch/cq_commands.hpp#L287

mo-tenstorrent committed:
#4984: 16B align mailbox's launch and profiler fields

mo-tenstorrent commented on tt_metal/hw/inc/dev_msgs.h:
```diff
@@ -234,13 +234,21 @@ struct dprint_buf_msg_t {
 
+
+// NOC aligment max from BH
+static constexpr uint32_t BH_NOC_ALIGNMENT = 64;
```
Wanted to be more descriptive that this is coming from Blackhole

pgkeller replied:
is MAX_NOC_ALIGNMENT the max across all supported arches?  If so, I disagree @mo-tenstorrent  and agree w/ @abhullar-tt

tt-rkim commented on .github/workflows/test-dispatch.yaml:
```diff
@@ -70,6 +71,7 @@ jobs:
       - uses: ./.github/actions/install-python-deps
       - name: Run pre/post regression tests in a loop
         run: |
+          echo ${{ inputs }}
```
Maybe a slightly more verbose log statement here

pgkeller committed:
#0: Fix ND hang in test_prefetcher w/ -spre -sdis

With split prefetcher/dispatcher the downstream component sets the MSB of the
credit a terminate signal.  The credit balancing wait waited for the credits to
match the expected credits but didn't account for the terminate signal.  This
would result in a hang if the downstream component finished before the upstream
omponent (which should be and was rare).
